### PR TITLE
refactor main to for ease of integration

### LINKED ISSUE #1813: wrap/refactor `get_extractor` and exception handling
see TODO in capa/main.py

### LINKED ISSUE #1821: move routines from main into dedicated modules
via #1820: 

Furthermore, does it make sense to break out any other logic from main? Here's whats left:

```
timing
set_vivisect_log_level
has_rule_with_namespace
is_internal_rule
is_file_limitation_rule
has_file_limitation
is_supported_format
is_supported_arch
get_arch
is_supported_os
get_os
get_meta_str
is_running_standalone
get_default_root
get_default_signatures
get_workspace
get_extractor
get_file_extractors
is_nursery_rule_path
collect_rule_file_paths
on_load_rule_default
get_rules
get_signatures
get_sample_analysis
collect_metadata
compute_dynamic_layout
compute_static_layout
compute_layout
install_common_args
handle_common_args
main
ida_main
ghidra_main
```

I think these things belong in main:

```
timing
set_vivisect_log_level
is_running_standalone
install_common_args
handle_common_args
main
ida_main
ghidra_main
```

proposal 1: move to existing capa.rules:

```
has_rule_with_namespace
is_internal_rule
is_file_limitation_rule
get_meta_str
is_nursery_rule_path
collect_rule_file_paths
on_load_rule_default
get_rules
```

proposal 2: move to `capa.capabilities`:

```
has_file_limitation
```

proposal 3: move to new module `capa.loader`:

```
has_file_limitation
is_supported_format
is_supported_arch
get_arch
is_supported_os
get_os
is_running_standalone
get_default_root
get_default_signatures
get_workspace
get_extractor
get_file_extractors
get_signatures
get_sample_analysis
collect_metadata
compute_dynamic_layout
compute_static_layout
compute_layout
```

Please brainstorm if these changes make intuitive sense to you. If so, then we can update this PR with the changes, paying particular attention to how the scripts change - we'd want these to become simpler/more readable.

_Originally posted by @williballenthin in https://github.com/mandiant/capa/issues/1820#issuecomment-1770344098_

### LINKED ISSUE #1945: OOM on certain files when using Python module
### Description

When using Capa as a library against certain files the process ends up just OOMing..

### Steps to Reproduce
Find a file that triggers it, I can provide a few samples if that helps.
https://www.virustotal.com/gui/file/f88e1f280af5b7ea78a8f0f59fba910e54e2eaeb2f34611a7e36f33b505d2784
https://www.virustotal.com/gui/file/f0fa2602f5b65dd91ec0eb7626556f6e9d07ed39065a574e05974a0bc3651017

1. Run bulk-process.py script against the files, eg;
```
python3 bulk-process.py -r capa-rules-6.1.0/ -s sigs/ -n1 --no-mp /tmp/samples
INFO:capa:successfully loaded 1061 rules
INFO:capa:computing capa results for: /tmp/samples/f88e1f280af5b7ea78a8f0f59fba910e54e2eaeb2f34611a7e36f33b505d2784.exe
Killed
```

**Expected behavior:**
Runs successfully against the file.

**Actual behavior:**
Process is killed by the oom killer.

### Versions
```
./capa --version
capa 6.1.0
```

pip list installed | grep -i flare
flare-capa        6.1.0   

python3 --version
Python 3.8.10

Using Ubuntu 20.04

### Additional Information
Interestingly, running the capa binary against the file works without any dramas.

### LINKED ISSUE #1946: rename args.sample to args.input
since the input file is not longer guaranteed to be an executable file to analyze (cape, freeze, result document), then we should rename the CLI argument from `sample` to `input` (?). we should do this in v7 since its a breaking change that goes along with the cape work.

### PR DESCRIPTION
**Pull request by williballenthin**
closes #1813
closes #1946 
closes #1821
closes #1945

This PR reorganizes the logic found in `main()` because it was getting very large and hairy. The key change is to pull out chunks of code into their own functions, typically labelled `*_from_cli()`, like `get_backend_from_cli()`. This localizes logic, which makes it easier for humans to read and make changes. It also means that `main()` becomes much more reasonably sized (less than 100 lines, about one screenful).

Typically we prefer to keep all input/output within main, calling library routines as necessary. This worked for a while, but `main` just became too large. The new "main routines" are expected to be used only within `main` functions, either capa main or related scripts. They reach directly into the argparse arguments, write to stdout/stderr, and can instruct the program to exit prematurely. These functions *should not* be invoked from library code, and this is documented.

Beyond copying code around, this PR refines our handling of the input file/format/backend. The logic for picking the format and backend is more consistent. Also, I've been careful to document that the input file is not necessarily the sample itself - cape/freeze/etc. inputs are not actually the sample.

While I was moving stuff around, I also addressed #1821, which moves a lot of potentially common routines from `capa.main` into the new namespace `capa.loader`, such as `is_supported_os`, `get_workspace`, `get_extractor`, etc. These routines don't depend directly on our CLI argument handling and can reasonably be used by other library code.

Here's an example of the changes needed to the freeze utility (they're basically 1 to 1):
image

But here's a more compelling example. The script needed to load rules and the input file, so we can replace a lot of the tedious exception handling with a consistent user interface (exceptions, status code, etc.):
image

TODO:
  - [x] port over script to use these new wrappers
  - [x] linters
  - [x] #1821 
  - [x] rename `*_from_args` to `*_from_cli`

### Checklist

- [ ] No CHANGELOG update needed

- [x] No new tests needed

- [x] No documentation update needed

### CODE REVIEW

File: capa/loader.py [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+      - elf
+      - shellcode 32-bit
+      - shellcode 64-bit
+      - auto
+
+    this creates and analyzes the workspace; however, it does *not* save the workspace.
+    this is the responsibility of the caller.
+    """
+
+    # lazy import enables us to not require viv if user wants SMDA, for example.
```
Comment (mr-tz) [inline capa/loader.py:138]:
```suggestion
    # lazy import enables us to not require viv if user wants another backend.
```

File: capa/main.py [hunk 2]
```diff
@@ -777,18 +221,18 @@ def install_common_args(parser, wanted=None):
     #
     # arguments that may be opted into:
     #
-    #   - sample
+    #   - input_file
     #   - format
     #   - os
     #   - rules
     #   - tag
     #
```
Comment (mr-tz) [inline capa/main.py:228]: duplication / addition to the above, let's combine the two

File: capa/main.py [hunk 3]
```diff
@@ -978,115 +455,133 @@ def handle_common_args(args):
...
+    raises:
+      ShouldExitError: if the program is invoked incorrectly and should exit.
+    """
+    if args.os:
+        return args.os
+
+    sample_path = get_sample_path_from_args(args, backend)
+    if sample_path is None:
+        return "unknown"
+    return capa.loader.get_os(sample_path)
```
Comment (mr-tz) [inline capa/main.py:569]: ah, neat

File: capa/main.py [hunk 4]
```diff
@@ -134,73 +123,13 @@ def set_vivisect_log_level(level):
...
-
-def is_running_standalone() -> bool:
-    """
-    are we running from a PyInstaller'd executable?
-    if so, then we'll be able to access `sys._MEIPASS` for the packaged resources.
-    """
-    return hasattr(sys, "frozen") and hasattr(sys, "_MEIPASS")
-
-
 def get_default_root() -> Path:
```
Comment (mr-tz) [inline capa/main.py:125]: should this function also go to helpers?

**Review by mr-tz — COMMENTED** [review]
looks great so far!!

File: capa/main.py (hunk 4 above)
Comment (williballenthin) [reply capa/main.py:125]: yeah i was a little torn by this. i think only capa.main will ever run from PyInstaller, so it seemed like `is_running_standalone` should be in main, which means that `get_default_root` can't be in capa.helpers (circular import).

Comment (mr-tz) [reply capa/main.py:125]: I see, let's document that in the source and keep it as proposed.

**Comment by williballenthin** [comment]
this *could* go into v7, though with beta1 already out, i'm not sure if its appropriate. i think these changes are solid, but if they don't pass code review easily, then we can push them off to v8.

**Review by mr-tz — COMMENTED** [review]
changes look good, it's quite a big change, but I'd prefer to get this into v7 now vs. doing a breaking v8 release next week

**Comment by mr-tz** [comment]
need to figure out what's going wrong with the `test_scripts` (`test_scripts[capa2yara.py-args0]`)

**Comment by williballenthin** [comment]
Tests passing again (locally, at least). I don't have any other pending changes. Merging this before v7 works for me.

**Review by mr-tz — APPROVED** [review]
LGTM, THANK YOU!

**Comment by mr-tz** [comment]
let's merge this?!

**Pull request merged**